#**Assignment_2: Intel Image Classification**

##**Problem Statement:**
####Perform experiments with images and check out the effect of batch processing, batch size, image spatial dimension etc. on the image data set. Find out the limitations of your machine in terms of processing the batch of images.

##**Dataset Description:**

This image data set focuses on natural scenes. It was created by Intel for an image
classification competition. Approximately 25,000 images are there in the dataset.
Images are grouped into categories such as buildings, forests, glaciers, mountains, seas,
and streets.
Dataset is divided into folders for training, testing, and prediction:

* 14,000 training images

* 3,000 validation images

* 7,000 test images

Link to the Dataset:

URL: https://www.kaggle.com/datasets/puneet6060/intel-image-classification

#**Methodology**:
In this model, input images of three different spatial dimensions are used and the model is trained by experimenting with two different batch sizes as mentioned below: 

* Input Image spatial dimension: 32, 64, 256.

* Batch sizes: 90, 150

#**Outcome**:
These experiments have been done to understand and know that how different spatial dimentions and batch processing affects the model, how much time it takes and what can be done to improve it further or make more efficient. 
 

In [3]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import Flatten
from keras.preprocessing import image
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D 
from sklearn.model_selection import train_test_split

In [4]:
!cp /content/kaggle.json ~/.kaggle/kaggle.json

In [5]:
!chmod 600 /root/.kaggle/kaggle.json

In [6]:
!kaggle datasets download -d puneet6060/intel-image-classification

intel-image-classification.zip: Skipping, found more recently modified local copy (use --force to force download)


In [7]:
import zipfile
zip_ref = zipfile.ZipFile('/content/intel-image-classification.zip', 'r')
zip_ref.extractall('/content')
zip_ref.close()

In [8]:
import glob
buildings = glob.glob('/content/seg_test/seg_test/buildings/*.*')
forest = glob.glob('/content/seg_test/seg_test/forest/*.*')
glacier = glob.glob('/content/seg_test/seg_test/glacier/*.*')
mountain = glob.glob('/content/seg_test/seg_test/moutain/*.*')
sea = glob.glob('/content/seg_test/seg_test/sea/*.*')
street = glob.glob('/content/seg_test/seg_test/street/*.*')

In [9]:
imgShape = [32,64,256]

##Image Dimention: 32 

In [10]:
data = []
labels = []

In [11]:
for i in buildings:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(0)
for i in forest:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(1)
for i in glacier:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in mountain:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(3)
for i in sea:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in street:    
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[0],imgShape[0]))
    image=np.array(image)
    data.append(image)
    labels.append(3)

In [12]:
data32 = np.array(data)
labels32 = np.array(labels)

In [13]:
x_train32, x_test32, y_train32, y_test32 = train_test_split(data32, labels32, test_size=0.2)

In [14]:
x_train32 = x_train32.astype('float32')
x_test32 = x_test32.astype('float32')
x_train32 /= 255
x_test32 /= 255

print(x_train32.shape[0], 'train samples')
print(x_test32.shape[0], 'test samples')

# Convert labels to categorical format
y_train32 = keras.utils.to_categorical(y_train32, 6)
y_test32 = keras.utils.to_categorical(y_test32, 6)

1980 train samples
495 test samples


In [15]:
model1 = Sequential()

model1.add(Conv2D(32, (5, 5), input_shape=(32,32,3), activation='relu', padding='same'))
model1.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model1.add(Conv2D(64, (5, 5), activation='relu', padding='same'))
model1.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model1.add(Conv2D(200, (5, 5), activation='relu', padding='same'))
model1.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model1.add(Flatten())


model1.add(Dense(300, activation='relu'))
model1.add(Dense(200, activation='relu'))
model1.add(Dropout(0.2))

model1.add(Dense(6, activation="softmax"))
model1.layers

##Image Dimention 32 and batch size 90

In [16]:
# Compile the model1
model1.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model1
history1= model1.fit(x_train32, y_train32, batch_size=90, epochs=15, validation_data=(x_test32, y_test32))


Epoch 1/15
22/22 [==============================] - 13s 29ms/step - loss: 1.2438 - accuracy: 0.4652 - val_loss: 0.9474 - val_accuracy: 0.6222
Epoch 2/15
22/22 [==============================] - 0s 13ms/step - loss: 0.8519 - accuracy: 0.6667 - val_loss: 0.7415 - val_accuracy: 0.7091
Epoch 3/15
22/22 [==============================] - 0s 12ms/step - loss: 0.7239 - accuracy: 0.7242 - val_loss: 0.6957 - val_accuracy: 0.7273
Epoch 4/15
22/22 [==============================] - 0s 11ms/step - loss: 0.6302 - accuracy: 0.7581 - val_loss: 0.6532 - val_accuracy: 0.7333
Epoch 5/15
22/22 [==============================] - 0s 12ms/step - loss: 0.5567 - accuracy: 0.7859 - val_loss: 0.6089 - val_accuracy: 0.7697
Epoch 6/15
22/22 [==============================] - 0s 10ms/step - loss: 0.5407 - accuracy: 0.7985 - val_loss: 0.6583 - val_accuracy: 0.7556
Epoch 7/15
22/22 [==============================] - 0s 12ms/step - loss: 0.4893 - accuracy: 0.8177 - val_loss: 0.6559 - val_accuracy: 0.7758
Epoch 8/15
2

In [17]:
test_loss32, test_acc32 = model1.evaluate(x_test32, y_test32)
print(f"Test accuracy with 32 input dimensions: {test_acc32}")

16/16 [==============================] - 0s 9ms/step - loss: 0.6129 - accuracy: 0.7879
Test accuracy with 32 input dimensions: 0.7878788113594055


##Image Dimention 32 and batch size 150

In [18]:
model1.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
history1_2= model1.fit(x_train32, y_train32, batch_size=150, epochs=15, validation_data=(x_test32, y_test32))


Epoch 1/15
14/14 [==============================] - 3s 42ms/step - loss: 0.6460 - accuracy: 0.8192 - val_loss: 0.7098 - val_accuracy: 0.7455
Epoch 2/15
14/14 [==============================] - 0s 14ms/step - loss: 0.2831 - accuracy: 0.9030 - val_loss: 0.7148 - val_accuracy: 0.7838
Epoch 3/15
14/14 [==============================] - 0s 15ms/step - loss: 0.1479 - accuracy: 0.9515 - val_loss: 0.6885 - val_accuracy: 0.8121
Epoch 4/15
14/14 [==============================] - 0s 15ms/step - loss: 0.0932 - accuracy: 0.9717 - val_loss: 0.7725 - val_accuracy: 0.8000
Epoch 5/15
14/14 [==============================] - 0s 15ms/step - loss: 0.0639 - accuracy: 0.9833 - val_loss: 0.7422 - val_accuracy: 0.8040
Epoch 6/15
14/14 [==============================] - 0s 14ms/step - loss: 0.0416 - accuracy: 0.9909 - val_loss: 0.8101 - val_accuracy: 0.8182
Epoch 7/15
14/14 [==============================] - 0s 14ms/step - loss: 0.0270 - accuracy: 0.9949 - val_loss: 0.8928 - val_accuracy: 0.8141
Epoch 8/15
14

In [19]:
test_loss32, test_acc32 = model1.evaluate(x_test32, y_test32)
print(f"Test accuracy with 32 input dimensions with 150 batch size : {test_acc32}")

16/16 [==============================] - 0s 5ms/step - loss: 1.2367 - accuracy: 0.8101
Test accuracy with 32 input dimensions with 150 batch size : 0.8101010322570801


##Image Dimention: 64

In [20]:
data = []
labels = []

In [21]:
for i in buildings:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(0)
for i in forest:   
    image=tf.keras.preprocessing.image.load_img(i,target_size=(imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(1)
for i in glacier:   
    image=tf.keras.preprocessing.image.load_img(i,target_size=(imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in mountain:   
    image=tf.keras.preprocessing.image.load_img(i,target_size=(imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(3)
for i in sea:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in street:    
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[1],imgShape[1]))
    image=np.array(image)
    data.append(image)
    labels.append(3)

In [22]:
data64 = np.array(data)
labels64 = np.array(labels)

In [23]:
x_train64, x_test64, y_train64, y_test64 = train_test_split(data64, labels64, test_size=0.2)

In [24]:
x_train64 = x_train64.astype('float32')
x_test64 = x_test64.astype('float32')
x_train64 /= 255
x_test64 /= 255

print(x_train64.shape[0], 'train samples')
print(x_test64.shape[0], 'test samples')

# Convert labels to categorical format
y_train64 = keras.utils.to_categorical(y_train64, 6)
y_test64 = keras.utils.to_categorical(y_test64, 6)

1980 train samples
495 test samples


In [25]:
model2 = Sequential()

model2.add(Conv2D(32, (5, 5), input_shape=(64,64,3), activation='relu', padding='same'))
model2.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model2.add(Conv2D(64, (5, 5), activation='relu', padding='same'))
model2.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model2.add(Conv2D(200, (5, 5), activation='relu', padding='same'))
model2.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model2.add(Flatten())


model2.add(Dense(300, activation='relu'))
model2.add(Dense(200, activation='relu'))
model2.add(Dropout(0.2))

model2.add(Dense(6, activation="softmax"))
model2.layers

##Image Dimention 64 and batch size 90

In [26]:
# Compile the model2
model2.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model2
history2 = model2.fit(x_train64, y_train64, batch_size=90, epochs=15, validation_data=(x_test64, y_test64))


Epoch 1/15
22/22 [==============================] - 3s 45ms/step - loss: 1.4045 - accuracy: 0.4343 - val_loss: 0.9434 - val_accuracy: 0.6162
Epoch 2/15
22/22 [==============================] - 1s 25ms/step - loss: 0.8569 - accuracy: 0.6485 - val_loss: 0.7959 - val_accuracy: 0.7273
Epoch 3/15
22/22 [==============================] - 1s 25ms/step - loss: 0.6827 - accuracy: 0.7318 - val_loss: 0.6731 - val_accuracy: 0.7556
Epoch 4/15
22/22 [==============================] - 1s 26ms/step - loss: 0.6171 - accuracy: 0.7586 - val_loss: 0.6434 - val_accuracy: 0.7677
Epoch 5/15
22/22 [==============================] - 1s 23ms/step - loss: 0.5896 - accuracy: 0.7773 - val_loss: 0.5729 - val_accuracy: 0.7899
Epoch 6/15
22/22 [==============================] - 1s 24ms/step - loss: 0.5603 - accuracy: 0.7889 - val_loss: 0.5692 - val_accuracy: 0.8141
Epoch 7/15
22/22 [==============================] - 0s 22ms/step - loss: 0.4702 - accuracy: 0.8263 - val_loss: 0.5933 - val_accuracy: 0.7899
Epoch 8/15
22

In [27]:
test_loss64, test_acc64 = model2.evaluate(x_test64, y_test64)
print(f"Test accuracy with 64 input dimensions: {test_acc64}")

16/16 [==============================] - 0s 9ms/step - loss: 0.7056 - accuracy: 0.8101
Test accuracy with 64 input dimensions: 0.8101010322570801


##Image Dimention 64 and batch size 150

In [28]:
model2.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
history2_2 = model2.fit(x_train64, y_train64, batch_size=150, epochs=15, validation_data=(x_test64, y_test64))


Epoch 1/15
14/14 [==============================] - 3s 64ms/step - loss: 0.4968 - accuracy: 0.8394 - val_loss: 0.6627 - val_accuracy: 0.7677
Epoch 2/15
14/14 [==============================] - 1s 36ms/step - loss: 0.1960 - accuracy: 0.9465 - val_loss: 0.6810 - val_accuracy: 0.8263
Epoch 3/15
14/14 [==============================] - 1s 36ms/step - loss: 0.0958 - accuracy: 0.9682 - val_loss: 0.7166 - val_accuracy: 0.8263
Epoch 4/15
14/14 [==============================] - 0s 35ms/step - loss: 0.0605 - accuracy: 0.9823 - val_loss: 0.7786 - val_accuracy: 0.8283
Epoch 5/15
14/14 [==============================] - 0s 36ms/step - loss: 0.0367 - accuracy: 0.9914 - val_loss: 0.8533 - val_accuracy: 0.8040
Epoch 6/15
14/14 [==============================] - 0s 34ms/step - loss: 0.0258 - accuracy: 0.9960 - val_loss: 0.9295 - val_accuracy: 0.8061
Epoch 7/15
14/14 [==============================] - 1s 36ms/step - loss: 0.0159 - accuracy: 0.9965 - val_loss: 0.9692 - val_accuracy: 0.8081
Epoch 8/15
14

In [29]:
test_loss64, test_acc64 = model2.evaluate(x_test64, y_test64)
print(f"Test accuracy with 64 input dimensions with 150 batch size: {test_acc64}")

16/16 [==============================] - 0s 5ms/step - loss: 1.3336 - accuracy: 0.8101
Test accuracy with 64 input dimensions with 150 batch size: 0.8101010322570801


##Image Dimention: 256 

In [30]:
data = []
labels = []

In [31]:
for i in buildings:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(0)
for i in forest:   
    image=tf.keras.preprocessing.image.load_img(i,target_size=(imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(1)
for i in glacier:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in mountain:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(3)
for i in sea:   
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(2)
for i in street:    
    image=tf.keras.preprocessing.image.load_img(i,target_size= (imgShape[2],imgShape[2]))
    image=np.array(image)
    data.append(image)
    labels.append(3)

In [32]:
data256 = np.array(data)
labels256 = np.array(labels)

In [33]:
x_train256, x_test256, y_train256, y_test256 = train_test_split(data256, labels256, test_size=0.2)

In [34]:
x_train256= x_train256.astype('float32')
x_test256 = x_test256.astype('float32')
x_train256 /= 255
x_test256 /= 255

print(x_train256.shape[0], 'train samples')
print(x_test256.shape[0], 'test samples')

# Convert labels to categorical format
y_train256 = keras.utils.to_categorical(y_train256, 6)
y_test256 = keras.utils.to_categorical(y_test256, 6)

1980 train samples
495 test samples


In [35]:
x_train256.shape

(1980, 256, 256, 3)

In [36]:
model3 = Sequential()

model3.add(Conv2D(32, (5, 5), input_shape=(256,256,3), activation='relu', padding='same'))
model3.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model3.add(Conv2D(64, (5, 5), activation='relu', padding='same'))
model3.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model3.add(Conv2D(200, (5, 5), activation='relu', padding='same'))
model3.add(MaxPooling2D(pool_size=(2, 2),strides=(2, 2)))

model3.add(Flatten())


model3.add(Dense(300, activation='relu'))
model3.add(Dense(200, activation='relu'))
model3.add(Dropout(0.2))

model3.add(Dense(6, activation="softmax"))
model3.layers

##Image Dimention 256 and batch size 90

In [37]:
# Compile the model3
model3.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model3
history3 = model3.fit(x_train256, y_train256, batch_size=90, epochs=15, validation_data=(x_test256, y_test256))


Epoch 1/15
22/22 [==============================] - 15s 416ms/step - loss: 1.6419 - accuracy: 0.4480 - val_loss: 0.8895 - val_accuracy: 0.6424
Epoch 2/15
22/22 [==============================] - 6s 274ms/step - loss: 0.8057 - accuracy: 0.6919 - val_loss: 0.6881 - val_accuracy: 0.7414
Epoch 3/15
22/22 [==============================] - 6s 276ms/step - loss: 0.6761 - accuracy: 0.7465 - val_loss: 0.6372 - val_accuracy: 0.7758
Epoch 4/15
22/22 [==============================] - 6s 277ms/step - loss: 0.5652 - accuracy: 0.7894 - val_loss: 0.5682 - val_accuracy: 0.8061
Epoch 5/15
22/22 [==============================] - 6s 281ms/step - loss: 0.4859 - accuracy: 0.8162 - val_loss: 0.5341 - val_accuracy: 0.8000
Epoch 6/15
22/22 [==============================] - 7s 339ms/step - loss: 0.3734 - accuracy: 0.8687 - val_loss: 0.5716 - val_accuracy: 0.8081
Epoch 7/15
22/22 [==============================] - 6s 270ms/step - loss: 0.2636 - accuracy: 0.9061 - val_loss: 0.5027 - val_accuracy: 0.8364
Epoch

In [87]:
test_loss256, test_acc256 = model3.evaluate(x_test256, y_test256)
print(f"Test accuracy with 256 input dimensions: {test_acc256}")

16/16 [==============================] - 2s 57ms/step - loss: 0.9467 - accuracy: 0.7838
Test accuracy with 256 input dimensions: 0.7838383913040161


##Image Dimention 256 and batch size 150

In [38]:
# Train the model3
history3_2 = model3.fit(x_train256, y_train256, batch_size=150, epochs=15, validation_data=(x_test256, y_test256))


Epoch 1/15
14/14 [==============================] - 14s 660ms/step - loss: 0.0109 - accuracy: 0.9975 - val_loss: 0.9633 - val_accuracy: 0.8141
Epoch 2/15
14/14 [==============================] - 6s 446ms/step - loss: 0.0056 - accuracy: 0.9985 - val_loss: 0.9715 - val_accuracy: 0.8283
Epoch 3/15
14/14 [==============================] - 6s 431ms/step - loss: 0.0038 - accuracy: 0.9990 - val_loss: 1.0982 - val_accuracy: 0.8222
Epoch 4/15
14/14 [==============================] - 6s 435ms/step - loss: 0.0032 - accuracy: 0.9995 - val_loss: 1.1108 - val_accuracy: 0.8444
Epoch 5/15
14/14 [==============================] - 6s 443ms/step - loss: 0.0015 - accuracy: 0.9995 - val_loss: 1.1062 - val_accuracy: 0.8364
Epoch 6/15
14/14 [==============================] - 6s 436ms/step - loss: 9.5090e-04 - accuracy: 1.0000 - val_loss: 1.1213 - val_accuracy: 0.8424
Epoch 7/15
14/14 [==============================] - 6s 437ms/step - loss: 6.8362e-04 - accuracy: 1.0000 - val_loss: 1.1738 - val_accuracy: 0.83

In [39]:
test_loss256, test_acc256 = model3.evaluate(x_test256, y_test256)
print(f"Test accuracy with 256 input dimensions with 150 batch size: {test_acc256}")

16/16 [==============================] - 2s 57ms/step - loss: 1.2829 - accuracy: 0.8303
Test accuracy with 256 input dimensions with 150 batch size: 0.8303030133247375


#**Observations and Conclusion**


####Test accuracy with 32 input and dimensions and 90 batch size: 0.79

####Test accuracy with 32 input dimensions and 150 batch size : 0.81

####Test accuracy with 64 input and dimensions and 90 batch size: 0.81

####Test accuracy with 64 input dimensions and 150 batch size : 0.81

####Test accuracy with 256 input and dimensions and 90 batch size: 0.78

####Test accuracy with 256 input dimensions and 150 batch size : 0.83

####While training the model I observed that 
* Larger batch size is giving more accuracy because a larger batch size can help to reduce overfitting, as each batch contains a more representative sample of the overall data distribution. But this is not true in all the cases.

* Larger batch size increased the run time. In our case the difference is not much.

* Larger input images require more memory to store the intermediate activations and gradients during training, which can be a limiting factor for GPUs with limited memory. This results in longer training time or the need to use smaller batch sizes.